### PR TITLE
New parallel replicas coordinator implementation

### LINKED ISSUE #57772: Improve locality of data distribution for parallel replicas
We already have a scheme in place that assigns parts to replicas based on consistent hash.
Currently it has the following problems:
* we assign parts as a whole, not taking part size (e.g. in marks) into account. because of that (let's omit hash function uniformity) distribution might be not really even
<details>
  <summary>Example</summary>

```
consistent_hash=0, assigned_parts[consistent_hash]=7, assigned_blocks[consistent_hash]=10248
consistent_hash=1, assigned_parts[consistent_hash]=4, assigned_blocks[consistent_hash]=5615
consistent_hash=2, assigned_parts[consistent_hash]=4, assigned_blocks[consistent_hash]=193
consistent_hash=3, assigned_parts[consistent_hash]=6, assigned_blocks[consistent_hash]=12355
consistent_hash=4, assigned_parts[consistent_hash]=6, assigned_blocks[consistent_hash]=13862
consistent_hash=5, assigned_parts[consistent_hash]=4, assigned_blocks[consistent_hash]=11261
consistent_hash=6, assigned_parts[consistent_hash]=5, assigned_blocks[consistent_hash]=6091
consistent_hash=7, assigned_parts[consistent_hash]=2, assigned_blocks[consistent_hash]=9810
consistent_hash=8, assigned_parts[consistent_hash]=2, assigned_blocks[consistent_hash]=1018	
consistent_hash=9, assigned_parts[consistent_hash]=4, assigned_blocks[consistent_hash]=10711
consistent_hash=10, assigned_parts[consistent_hash]=5, assigned_blocks[consistent_hash]=6034
consistent_hash=11, assigned_parts[consistent_hash]=2, assigned_blocks[consistent_hash]=5024
consistent_hash=12, assigned_parts[consistent_hash]=3, assigned_blocks[consistent_hash]=6235
consistent_hash=13, assigned_parts[consistent_hash]=2, assigned_blocks[consistent_hash]=5572
consistent_hash=14, assigned_parts[consistent_hash]=7, assigned_blocks[consistent_hash]=16289
consistent_hash=15, assigned_parts[consistent_hash]=4, assigned_blocks[consistent_hash]=5396
```

</details>

* each query consists of two stages: receiving announcements from all the replicas and then the main phase when all parts were assigned taking into account unavailable nodes. if the first stage takes a long time some replicas will already exhaust their assignments and start to read from other replica's parts taking whatever came to their hands (without consistent hashing)
* make sure replicas numbers are consistent across cluster #57800

---

would be good to have a more accurate distribution. would be even better if will provide a decent cache hit for queries with different filters.

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Default coordinator for parallel replicas is rewritten for better cache locality (same mark ranges are almost always assigned to the same replicas). Consistent hashing is used also during work stealing, so better tail latency is expected.
It has been tested for linear scalability on a hundred of replicas.

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 3866f26cad7b53c509297d0ed8f007af6ae98b80 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57968/3866f26cad7b53c509297d0ed8f007af6ae98b80/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7239641699">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57968/3866f26cad7b53c509297d0ed8f007af6ae98b80/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by devcrafter** [comment]
Still need to check details but some notes:
(1) Specific tests for previous coordinator are removed. AFAIU, current tests cover only cases when part set is the same on all replicas, so there are no tests for stealing behavior. Let's add test(s)
(2) Related to (1). The current way of choosing a working set is non-deterministic. It'd be easier to test different cases if the working set would be taken from the initiator. We've discussed it, but I'm not really convinced that it'd be much more hassle to implement
(3) with current PR, the default coordinator is changed. `InOrder` coordinators left the same for now. I guess for tables with few big parts, new implementation can outperform `InOrder` implementation. So, at least it'd be nice to ensure that it's possible to switch to the default coordinator by disabling in order optimization settings

File: src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp [hunk 1]
```diff
@@ -98,165 +175,216 @@ using Parts = std::set<Part>;
...
+    /// observed along the way to what node they belong to.
+    /// Ranges in this queue might belong to a part that the given replica cannot read from - the corresponding check happens later.
+    /// TODO: consider making it bounded in size
+    std::vector<std::multiset<RangesInDataPartDescription, BiggerPartsFirst>> distribution_by_hash_queue;
 
-    ParallelReadResponse handleRequest(ParallelReadRequest request) override;
-    void handleInitialAllRangesAnnouncement(InitialAllRangesAnnouncement announcement) override;
-    void markReplicaAsUnavailable(size_t replica_number) override;
+    /// For some ranges their owner and stealer (by consistent hash) cannot read from the given part at all. So this range have to be stolen anyway.
+    /// Also if those replicas unavailable during the query - segments also will end up here.
```
Comment (devcrafter) [inline src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:272]: The comment is misleading. If a part is in set, then a replicas announced it. If the replica becomes unavailable after announcement - currently, the query fails.

File: src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp [hunk 2]
```diff
@@ -269,116 +397,354 @@ void DefaultCoordinator::finalizeReadingState()
...
-            part->description.ranges.pop_front();
         }
+
+        if (!result.ranges.empty())
+            response.description.push_back(std::move(result));
+    }
+}
+
+void DefaultCoordinator::selectPartsAndRanges(
+    size_t replica_num, ScanMode scan_mode, size_t min_number_of_marks, size_t & current_marks_amount, ParallelReadResponse & response)
```
Comment (devcrafter) [inline src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:631]: Let's separate protocol and implementation. There is no need to pass the response as a parameter here, only `response.description` (`RangesInDataPartsDescription`) is used through the calls.

File: src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp [hunk 3]
```diff
@@ -269,116 +397,354 @@ void DefaultCoordinator::finalizeReadingState()
...
     if (sent_initial_requests == replicas_count)
-        finalizeReadingState();
+        setProgressCallback();
+
+    /// Sift the queue to move out all invisible segments
+    for (const auto & segment : distribution_by_hash_queue[replica_num])
+    {
+        if (!part_visibility[segment.info.getPartNameV1()].contains(replica_num))
+        {
+            chassert(segment.ranges.size() == 1);
```
Comment (devcrafter) [inline src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:428]: Don't get it, how it's guaranteed?

File: src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp [hunk 4]
```diff
@@ -269,116 +397,354 @@ void DefaultCoordinator::finalizeReadingState()
...
+            enqueueToStealerOrStealingQueue(segment.info, segment.ranges.front());
+        }
+    }
 }
 
-void DefaultCoordinator::selectPartsAndRanges(const PartRefs & container, size_t replica_num, size_t min_number_of_marks, size_t & current_mark_size, ParallelReadResponse & response) const
+void DefaultCoordinator::tryToTakeFromDistributionQueue(
+    size_t replica_num, ScanMode, size_t min_number_of_marks, size_t & current_marks_amount, ParallelReadResponse & response)
```
Comment (devcrafter) [inline src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:435]: `scan_mode` is not used, please remove

**Review by devcrafter — COMMENTED** [review]
In general, It'd make sense to check that the working set is completely covered by ranges sent to replicas. Probably, it can be disabled in release

**Comment by nickitat** [comment]
> In general, It'd make sense to check that the working set is completely covered by ranges sent to replicas. Probably, it can be disabled in release

maybe, but it won't be always true. e.g. when we have `limit`.

**Comment by nickitat** [comment]
`Stress test (debug)` - #58378
`Upgrade check (debug)` - #57893

**Pull request merged**